Sources/client/Socket.swift

```diff
@@ -525,6 +519,39 @@ public class Socket {
     let msg = items.map( { return String(describing: $0) } ).joined(separator: ", ")
     self.logger?("SwiftPhoenixClient: \(msg)")
   }
+
+  /// - parameter endPoint: a base URL in string form
+  /// - parameter params: an optional function that returns an optional parameter dictionary for the connection stage
+  /// - return: a fully qualified URL built fron endPoint and params
+  internal static func buildUrl(endPoint: String, params: PayloadGetterFunction?) -> URL {
```

Comment:
I don't like making this static. I think the best move to do here is delete `var endpointUrl URL` and instead convert it to a internal function that builds and returns the URL like this function is doing.

similar to how the [js libary does it](https://github.com/phoenixframework/phoenix/blob/master/assets/js/phoenix.js#L798)

Reply:
I made it static as it is meant to have no state, input -> output. But sure, I could make it an instance function instead and remove the var if that's what you want.

Reply:
To preserve the `var websocketProtocol: String` I'll have to rebuild that url each time, though. Is this acceptable?

Reply:
1. Previously, we stored the `endpointUrl` as `public let endpointUrl: URL` and it was set at initialization. Instead, we should move it to be a class function which will reevaluate the endpointUrl at the time of calling

Declaration
```Swift
func endpointUrl(): URL {
    guard let url = /* build url here */ else { fatalError("malformed")
    return url
}

func connect() {
    self.connection = self.transport(self.endpointUrl())
    ...
}
```

This allows a client to call `.endpointUrl()` on their socket instance and it will always return the URL associated with that `Socket` instance

2. I think we should create a `private let endpointComponents: URLComponents` which would be set to `urlComponents` in the initializer and then in `websocketProtocol()` we can still return `endpointUrlComponents.scheme`. So replacing the `private let endpointURL()` (which has now become a public function) with a private variable.

And then when we build the `endpointUrl()`, we do something like

```Swift
self.urlComponents.queryItems = ...
```

This avoid rebuilding the endpointUrl inside of the `websocketProtocol()` function and removes the status builder.

Reply:
I'm just going in circles here now. I cannot keep `public var endPointUrl: URL` as I cannot call an instance method before it is fully initialised, and that's why I made it static to begin with, which did not bother me because it is a very simple deterministic function.

Trying to avoid that I made it a computed property, but that is not a good idea either simply because of how params can change, thus it would make it possible for  `var endPointURL: URL { ... }` to return different URLs despite not having changed the connections state, meaning it does not reflect what the socket is actually connected with. Meaning we would want to save it and not recompute it, which to solve the init-race would require a static method.

Or do you have any other ideas to get around this? I'd also add `private(set) var ...` to it just to make sure it can't be overridden by the consumer.